File: src/p0008/python/solution.py
class Solution:
    def myAtoi(self, s: str) -> int:

        i = 0

        while i < len(s) and s[i] == ' ':
            i += 1

        sign = 1
        if i < len(s):
            if s[i] == '-':
                sign = -1
                i += 1
            elif s[i] == '+':
                i += 1

        y = 0
        while i < len(s) and s[i].isdigit():
            if y > 0x7fffffff // 10 or (y == 0x7fffffff // 10 and s[i] > '7'):
                return 0x7fffffff if sign > 0 else -0x80000000
            y, i = y * 10 + int(s[i]), i + 1

        return sign * y


if __name__ == '__main__':
    sol = Solution()
    print(sol.myAtoi("2147483648"))

File: src/p0008/cpp/solution.cpp
#include <iostream>
#include <string>

using namespace std;


class Solution {
public:
    int myAtoi(string &s) {

        int i = 0;

        // Consume whitespaces.
        while (i < s.length() && s[i] == ' ') i++;

        // Consume sign.
        int32_t sign = 1;
        if (i >= s.length()) return 0;
        switch (s[i]) {
            case '-': sign = -1;
            case '+': i++; break;
        }

        // Consume value.
        int32_t y = 0;
        while (i < s.length() && isdigit(s[i])) {
            if (y > INT_MAX / 10 || (y == INT_MAX / 10 && s[i] > '7')) {
                return (sign > 0) ? INT_MAX : INT_MIN;
            }
            y = y * 10 + (s[i] - '0'); i++;
        }

        return sign * y;
    }
};

int main() {
    Solution sol;
    string s("-91283472332");
    cout << sol.myAtoi(s) << endl;
}